### DISCUSSION TITLE
How do I run jobs in a specific millisecond?

### DISCUSSION THREAD

**Discussion by mojtabamn22** [quartznet/quartznet#976] [General]
I want jobs to start running in certain milliseconds.
For example, start at 2:00:00.200, 2:00:00.400, 2:00:00.600, and so on. It may take a few minutes to complete these tasks, it is important to me that they are performed at the specified time (milliseconds).
I used the following code to do this but it did not run at the specified time

```csharp
StdSchedulerFactory factory = new StdSchedulerFactory();
IScheduler scheduler = await factory.GetScheduler();
await scheduler.Start();
IJobDetail job = JobBuilder.Create<HelloJob>()
    .WithIdentity("job1", "group1")
    .Build();
ITrigger trigger = TriggerBuilder.Create()
    .WithIdentity("trigger1", "group1")
    .StartAt(DateBuilder.DateOf(2, 0, 0))
    .WithSimpleSchedule(x => x
                        .WithInterval(TimeSpan.FromMilliseconds(200))
                        .WithRepeatCount(5))
    .Build();
await scheduler.ScheduleJob(job, trigger);
await Task.Delay(TimeSpan.FromSeconds(60));
await scheduler.Shutdown();
```

**Comment by lahma** [comment]
If you are building a real time system I'm afraid that a scheduling library will have hard time to meet your expectations. There's always some overhead which will cause jobs to run some milliseconds late.

Can you elaborate how the code did not run at specified time? Was it off by couple milliseconds or didn't it run at all?

**Reply by mojtabamn22** [reply]
> If you are building a real time system I'm afraid that a scheduling library will have hard time to meet your expectations. There's always some overhead which will cause jobs to run some milliseconds late.
> 
> Can you elaborate how the code did not run at specified time? Was it off by couple milliseconds or didn't it run at all?

I use these codes but instead of running at 18:58: 50.900 it run at "Greetings from HelloJob! 06: 58: 51.533"

```csharp
StdSchedulerFactory factory = new StdSchedulerFactory();
IScheduler scheduler = await factory.GetScheduler();

// and start it off
await scheduler.Start();

// define the job and tie it to our HelloJob class
IJobDetail job = JobBuilder.Create<HelloJob>()
    .WithIdentity("job1", "group1")
    .Build();

// Trigger the job to run now, and then repeat every 10 seconds
ITrigger trigger = TriggerBuilder.Create()
    .WithIdentity("trigger1", "group1")
    .StartAt(DateTime.Today.Add(new TimeSpan(0, 18, 58, 50, 900)))
    //.WithSimpleSchedule(x => x
                        //.WithInterval(TimeSpan.FromMilliseconds(200))
                       // .WithRepeatCount(5))
    .Build();

// Tell quartz to schedule the job using our trigger
await scheduler.ScheduleJob(job, trigger);

// some sleep to show what's happening
await Task.Delay(TimeSpan.FromSeconds(60));

// and last shut down the scheduler when you are ready to close your program
await scheduler.Shutdown();
```

**Reply by lahma** [reply]
When I run on my own computer with schedule `StartAt(DateTime.Today.Add(new TimeSpan(0, 19, 22, 50, 900)))` I get output `2020-09-29 19:22:50.954` so there's some penalty from infrastructure.

**Reply by mojtabamn22** [reply]
> When I run on my own computer with schedule `StartAt(DateTime.Today.Add(new TimeSpan(0, 19, 22, 50, 900)))` I get output `2020-09-29 19:22:50.954` so there's some penalty from infrastructure.

I left Visual Studio and tried again, the result was better. But it was 40 milliseconds late. Can you do several jobs at a time, for example 5:00:00.100, 5:00:00.150, 5:00:00.200, etc. (50 milliseconds apart)

**Reply by lahma** [reply]
Infrastructure has some penalty what you've experienced. Some of it could probably be optimized away.

It's multi-threaded (by default 10 workers) so you can schedule how you want by having multiple triggers.

**Comment by mojtabamn22** [comment]
@lahma 
I use the following code to add rows to the datagrid, but nothing happens
```csharp
public class HelloJob : IJob
{
    Form1 frm = new Form1();
    public async Task Execute(IJobExecutionContext context)
    {
        //await Console.Out.WriteLineAsync("Greetings from HelloJob!" + DateTime.Now.ToString("hh:mm:ss.fff"));
        await Task.Run(() =>
        {
            frm.dataGridView1.Rows.Add(DateTime.Now.ToString("hh:mm:ss.fff"));
        });
    }
}

private async void button1_Click(object sender, EventArgs e)
{
    StdSchedulerFactory factory = new StdSchedulerFactory();
    IScheduler scheduler = await factory.GetScheduler();

    // and start it off
    await scheduler.Start();

    // define the job and tie it to our HelloJob class
    IJobDetail job = JobBuilder.Create<HelloJob>()
        .WithIdentity("job1", "group1")
        .Build();

    // Trigger the job to run now, and then repeat every 10 seconds
    ITrigger trigger = TriggerBuilder.Create()
        .WithIdentity("trigger1", "group1")
        .StartAt(DateTime.Today.Add(new TimeSpan(0, 18, 39, 20, 900)))
        .WithSimpleSchedule(x => x
            .WithInterval(TimeSpan.FromMilliseconds(50))
            .WithRepeatCount(30))
        .Build();

    // Tell quartz to schedule the job using our trigger
    await scheduler.ScheduleJob(job, trigger);

    // some sleep to show what's happening
    await Task.Delay(TimeSpan.FromSeconds(30));

    // and last shut down the scheduler when you are ready to close your program
    await scheduler.Shutdown();

    Form1 frm = new Form1();
    frm.textBox1.Text = "Finish";
}
```

**Reply by lahma** [reply]
This is now taking a whole different tangent. I don't think that you can work with windows forms like that (creating in outside job and not even showing it).

**Reply by mojtabamn22** [reply]
> This is now taking a whole different tangent. I don't think that you can work with windows forms like that (creating in outside job and not even showing it).

How can I work with quartz in Windows Form? Is there no way to use the in Windows form?

**Reply by lahma** [reply]
I don't the answers to this. UI threading needs to be handled properly as jobs are running outside of main UI thread.
